# Import Packages

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from scipy.spatial.distance import pdist, squareform


# display all columns
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows', 60)

pd.options.display.max_rows = 60

/Users/angela/opt/anaconda3/lib/python3.7/site-packages/sklearn/feature_extraction/image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):


# Import Behold files

In [2]:
# import excel file 1
df1 = pd.read_csv('1_behold_brands USC.csv')
# import excel file 2
df2 = pd.read_excel('2_Behold+product+data+04262021.xlsx')
# import excel file 3
df3 = pd.read_csv('3_outfit_combinations USC.csv')
# import excel file 4
df4 = pd.read_csv('4_usc_additional_tags USC.csv')

In [3]:
print('file 1 shape: ' + str(df1.shape))
print('file 2 shape: ' + str(df2.shape))
print('file 3 shape: ' + str(df3.shape))
print('file 4 shape: ' + str(df4.shape))

file 1 shape: (162, 10)
file 2 shape: (61355, 11)
file 3 shape: (5291, 5)
file 4 shape: (97420, 4)


## File 1

We found that `quote`, `quote_attribute`, `intro`, `life_style_copy`, and `short_bio` are not neccessayrily needed for this project analysis, thus, we have decided to leave them aside by dropping of those columns.



In [4]:
df1 = df1.drop(['quote','quote_attribute','intro','lifestyle_copy','short_bio'], axis=1)
df1

brand_id            brand                 brand_value  \
0    01ESKR0CH2KYC7KBNTN0S38EQA  Mari Giudicelli  Handmade / Artisan Crafted   
1    01ESKR0CH2KYC7KBNTN0S38EQA  Mari Giudicelli                 Sustainable   
2    01ESKR0CH2KYC7KBNTN0S38EQA  Mari Giudicelli                 Women Owned   
3    01ESVD80YYAJQEQPADZ7HTHZXA             Warm                         NaN   
4    01EFJFZ329YVC8SK87R2BNJGV0            Vaara                    Emerging   
..                          ...              ...                         ...   
157  01EFJFYWNREWCT3Q4J9VTJ3ZSS    Kin the Label  Handmade / Artisan Crafted   
158  01EFJFYWNREWCT3Q4J9VTJ3ZSS    Kin the Label                 Women Owned   
159  01EFJFYWNREWCT3Q4J9VTJ3ZSS    Kin the Label                    Emerging   
160  01EH3BTNP1JQNMQYFMQWXJX9SM         Triarchy                 Sustainable   
161  01EH3BTNP1JQNMQYFMQWXJX9SM         Triarchy                    Emerging   

                                                   bio  \
0                                                  NaN   
1                                                  NaN   
2                                                  NaN   
3                                                  NaN   
4                                                  NaN   
..                                                 ...   
157  KIN THE LABEL is the creative vision of twin s...   
158  KIN THE LABEL is the creative vision of twin s...   
159  KIN THE LABEL is the creative vision of twin s...   
160  Triarchy leads the way in sustainable denim ma...   
161  Triarchy leads the way in sustainable denim ma...   

                                           listing_bio  
0    Behold Mari Giudicelli! This Brazilian shoe de...  
1    Behold Mari Giudicelli! This Brazilian shoe de...  
2    Behold Mari Giudicelli! This Brazilian shoe de...  
3    Behold Warm! Founded by specialty store owner,...  
4                                                  NaN  
..                                                 ...  
157  Behold Kin the Label! Top off your look with a...  
158  Behold Kin the Label! Top off your look with a...  
159  Behold Kin the Label! Top off your look with a...  
160  Behold Triarchy! Founded by three siblings, th...  
161  Behold Triarchy! Founded by three siblings, th...  

[162 rows x 5 columns]

## File 2

We cannot use `brand_canonical_url` because it is generated after the products are onboarded. Plus, `created_at` is the date stamp related to  `brand_canonical_url`. Thus, we don't need those 2 columns.

`product_active` is referred to whether the product is current active or inactive. We may need this column in Part 2 of this project as if the product is not active, this product shouldn't be used in the outfit reccommendation. We may add it back in Part 2

In [5]:
# drop brand_canonical_url, created_at, product_active  col
df2 = df2.drop(['brand_canonical_url','created_at','product_active'], axis=1)
df2

product_id           brand        brand_category  \
0      01EX0PN4J9WRNZH5F93YEX6QAF             Two               Unknown   
1      01F0C4SKZV6YXS3265JMC39NXW  Collina Strada               Unknown   
2      01EY4Y1BW8VZW51BWG5VZY82XW         Cariuma               Unknown   
3      01EY50E27A0P5V6KCW01XPDB43         Cariuma               Unknown   
4      01EY6DWHC2W5HPNEGXKEJ4A1CX         Cariuma               Unknown   
...                           ...             ...                   ...   
61350  01EYB5ERGYPFNGM6C9QK7Q9EV0         Nomasei          SandalsSales   
61351  01EHWTBFP368Q035FW95TRJDAY         Nomasei  Flat sandalsArchives   
61352  01EHWTCFTPPSCW10D4XBQZF28H         Nomasei  Flat sandalsArchives   
61353  01EYB5B5FH7JESXF82ZEMVXZMS         Nomasei          SandalsSales   
61354  01EHWTF4VNZ6FPF2836A0CFEEG         Nomasei  Flat sandalsArchives   

                                                    name  \
0                 Khadi Stripe Shirt-our signature shirt   
1          RUFFLE MARKET DRESS LOOPY PINK SISTINE TOMATO   
2                 IBI Slip On Raw Red Knit Sneaker Women   
3                   IBI Slip On Black Knit Sneaker Women   
4      CATIBA PRO Skate Black Suede and Canvas Contra...   
...                                                  ...   
61350                BowVida Mule in Black Suede Kidskin   
61351       Sandale Vida Mule In Tangerine Suede Kidskin   
61352              BowVida Mule in Fuschia Suede Kidskin   
61353              Vida Mule in Silver Metalized Leather   
61354           BowVida Mule in Bronze Metalized Leather   

                                                 details  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
61350  Feminine flat mules.\nSquare shape, V line on ...   
61351  Feminine flat mules.\nSquare shape, V line on ...   
61352  Feminine flat mules.\nSquare shape, V line on ...   
61353  Feminine flat mules.\nSquare shape, V line on ...   
61354  Feminine flat mules.\nSquare shape, V line on ...   

                                             description  \
0      Our signature khadi shirt\navailable in black ...   
1      Mid-length dress with ruffles and adjustable s...   
2                 IBI Slip On Raw Red Knit Sneaker Women   
3                   IBI Slip On Black Knit Sneaker Women   
4                                                    NaN   
...                                                  ...   
61350  The flat BowVida mule in black suede is the id...   
61351  The flat Vida mule in tangerine suede is the c...   
61352  The flat BowVida mule in fuschia suede is the ...   
61353  The flat Vida mule in silver metalized leather...   
61354  The flat BowVida mule in bronze leather is the...   

                                       brand_description  \
0      Our signature khadi shirt\n\navailable in blac...   
1      Mid-length dress with ruffles and adjustable s...   
2                 IBI Slip On Raw Red Knit Sneaker Women   
3                   IBI Slip On Black Knit Sneaker Women   
4                                                    NaN   
...                                                  ...   
61350  The flat BowVida mule in black suede is the id...   
61351  The flat Vida mule in tangerine suede is the c...   
61352  The flat BowVida mule in fuschia suede is the ...   
61353  The flat Vida mule in silver metalized leather...   
61354  The flat BowVida mule in bronze leather is the...   

                                              brand_name  
0                 Khadi Stripe Shirt-our signature shirt  
1          RUFFLE MARKET DRESS LOOPY PINK SISTINE TOMATO  
2                 IBI Slip On Raw Red Knit Sneaker Women  
3 

## File 3

There are some duplicate rows which have `product_id` with name = "`#NAME?`" and all else columns values are identical. Thus, we have decided to get rid of these rows to avoid repeated data which may lead to inaccuracy analysis 

In [6]:
## Full steps to identify duplicate product_id with name = "#NAME?" and delete all rows
# df3 =pd.read_csv('3_outfit_combinations USC.csv')
dup = df3[df3['product_id'].duplicated(keep = False)] ## keep all duplicates
# filter all duplicates for those with name '#NAME?'
dup_delete = dup[dup['product_full_name'] == '#NAME?']
    
#compare shape of original data frame (before and after)
print('original_df3_shape :' ,df3.shape)

# delete duplicates with faulty name from original population data frame
df3 = df3.drop(dup_delete.index)

# print updated shape of file
print('updated_df3_shape :', df3.shape)

#verify no observations in original data frame with '#NAME?'
df3[df3['product_full_name']=='#NAME?'].shape


original_df3_shape : (5291, 5)
updated_df3_shape : (5252, 5)


(0, 5)

We are able to optain new `df3` dataset now. Below is all current data shape

In [7]:
print('file 1 shape: ' + str(df1.shape))
print('file 2 shape: ' + str(df2.shape))
print('file 3 shape: ' + str(df3.shape))
print('file 4 shape: ' + str(df4.shape))

file 1 shape: (162, 5)
file 2 shape: (61355, 8)
file 3 shape: (5252, 5)
file 4 shape: (97420, 4)


# Merge files

## Merge `df3` and `df4`

Before we start, we want to make sure if all product_id of one dataset is the subset of the others

In [8]:
## check product_id matches between df3 and df4

df3ProdList = df3['product_id'].tolist()

df4ProdList = df4['product_id'].tolist()

print('shape df4: ', df4.shape)

print('shape df3: ', df3.shape)

shape df4:  (97420, 4)
shape df3:  (5252, 5)


In [9]:
#df4[df4['product_id'].isin(df3ProdList)]

df3[df3['product_id'].isin(df4ProdList)]

outfit_id                  product_id outfit_item_type  \
0     01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1           bottom   
1     01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T              top   
2     01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2S5T9W793F4CY41HE       accessory1   
3     01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2ZFDYRYY5TRQZJTBD             shoe   
4     01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1           bottom   
...                          ...                         ...              ...   
5286  01E6MC52DNZE5AZ9MPZ2CYCXW9  01E5ZYHZA7186DVWEJ99Q4D2PM       accessory2   
5287  01E6MC52DPCTAAQVYNSKA7PER7  01E2P0SJSKFKNQJ5SVQ8MD1JZT             shoe   
5288  01E6MC52DPCTAAQVYNSKA7PER7  01E4RW25Y8ZF6WKZRE50Y6SKH5         onepiece   
5289  01E6MC52DPCTAAQVYNSKA7PER7  01E5ZS3R9JD696YWGK9NSG56E1       accessory1   
5290  01E6MC52DPCTAAQVYNSKA7PER7  01E5ZYHZA7186DVWEJ99Q4D2PM       accessory2   

                    brand                         product_full_name  
0           Eileen Fisher                           Slim Knit Skirt  
1           Eileen Fisher                        Rib Mock Neck Tank  
2     kate spade new york            medium margaux leather satchel  
3              Tory Burch                 Penelope Mid Cap Toe Pump  
4           Eileen Fisher                           Slim Knit Skirt  
...                   ...                                       ...  
5286          Sam Edelman  65mm Gradient Oversize Square Sunglasses  
5287          DR. MARTENS               Fenimore Triple Buckle Boot  
5288           Nili Lotan                              Pamela Dress  
5289       Mansur Gavriel              Leather Circle Crossbody Bag  
5290          Sam Edelman  65mm Gradient Oversize Square Sunglasses  

[5252 rows x 5 columns]

Now, after verifying, we can go ahead and merge `df3` with `df4` on `product_id` using `outer join` to keep all the columns from both datasets

In [10]:
# merge dataset tags and outfit_combination using outer join on product_id
df34 = pd.merge(df3,df4, on ='product_id', how = 'outer')
df34

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
211126                         NaN  01E2KXWWY6JAGENED7P62090NP   
211127                         NaN  01E2KXWWY6JAGENED7P62090NP   
211128                         NaN  01E2KXWWY6JAGENED7P62090NP   
211129                         NaN  01DPD3EY7GVQ19WQR7P0K4BAZA   
211130                         NaN  01DVRGX5BT8VARMJB9THRM5MWG   

       outfit_item_type          brand   product_full_name  \
0                bottom  Eileen Fisher     Slim Knit Skirt   
1                bottom  Eileen Fisher     Slim Knit Skirt   
2                bottom  Eileen Fisher     Slim Knit Skirt   
3                bottom  Eileen Fisher     Slim Knit Skirt   
4                   top  Eileen Fisher  Rib Mock Neck Tank   
...                 ...            ...                 ...   
211126              NaN            NaN                 NaN   
211127              NaN            NaN                 NaN   
211128              NaN            NaN                 NaN   
211129              NaN            NaN                 NaN   
211130              NaN            NaN                 NaN   

                  product_color_id     attribute_name      attribute_value  
0       01DMBRYVA2C7BQN0X7MNNYXBHW       primarycolor               beiges  
1       01DMBRYVA2C7BQN0X7MNNYXBHW           category               bottom  
2       01DMBRYVA2C7BQN0X7MNNYXBHW       primarycolor               beiges  
3       01DMBRYVA2C7BQN0X7MNNYXBHW           category               bottom  
4       01DMBRYVA23E5A8GBES8QJN3HF           category  blazerscoatsjackets  
...                            ...                ...                  ...  
211126  01E2KXWWYGMCVTNKR9MH336WRV  subcategorybottom        pantsleggings  
211127  01E2KXWWYGMCVTNKR9MH336WRV       drycleanonly                  yes  
211128  01E2KXWWYGMCVTNKR9MH336WRV                fit       fittedtailored  
211129  01DPD3EY7NYEF766HGD9M2J9W6           category                  top  
211130  01DVRGX5C22Z5WEJXP2X1SJ7J6           category                  top  

[211131 rows x 8 columns]

## Merge `df34` with `df2`

As same as above merging, we use `outer join` to merge `df34` with `df2` on `product_id`

In [11]:
# outer join file_3_4 and product_data_2 (merge on product_id)
df234 = pd.merge(df34,df2, on ='product_id', how = 'outer')
df234

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x   product_full_name  \
0                bottom  Eileen Fisher     Slim Knit Skirt   
1                bottom  Eileen Fisher     Slim Knit Skirt   
2                bottom  Eileen Fisher     Slim Knit Skirt   
3                bottom  Eileen Fisher     Slim Knit Skirt   
4                   top  Eileen Fisher  Rib Mock Neck Tank   
...                 ...            ...                 ...   
268701              NaN            NaN                 NaN   
268702              NaN            NaN                 NaN   
268703              NaN            NaN                 NaN   
268704              NaN            NaN                 NaN   
268705              NaN            NaN                 NaN   

                  product_color_id attribute_name      attribute_value  \
0       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
1       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
2       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
3       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
4       01DMBRYVA23E5A8GBES8QJN3HF       category  blazerscoatsjackets   
...                            ...            ...                  ...   
268701                         NaN            NaN                  NaN   
268702                         NaN            NaN                  NaN   
268703                         NaN            NaN                  NaN   
268704                         NaN            NaN                  NaN   
268705                         NaN            NaN                  NaN   

        brand_y        brand_category  \
0           NaN                   NaN   
1           NaN                   NaN   
2           NaN                   NaN   
3           NaN                   NaN   
4           NaN                   NaN   
...         ...                   ...   
268701  Nomasei          SandalsSales   
268702  Nomasei  Flat sandalsArchives   
268703  Nomasei  Flat sandalsArchives   
268704  Nomasei          SandalsSales   
268705  Nomasei  Flat sandalsArchives   

                                                name  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   
...                                              ...   
268701           BowVida Mule in Black Suede Kidskin   
268702  Sandale Vida Mule In Tangerine Suede Kidskin   
268703         BowVida Mule in Fuschia Suede Kidskin   
268704         Vida Mule in Silver Metalized Leather   
268705      BowVida Mule in Bronze Metalized Leather   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
26870

## Export to csv file

We now have a merge file `df234`. Let's save it under csv file for later use!

In [12]:
df234.to_csv('df234.csv')

# Combine columns `name`, `brand_name`, and `product_full_name`

## Columns `name` vs `brand_name` vs `product_full_name`

When looking briefly on these 3 columns, we have found that there are major similarities among those 3, thus, we have decided to look into those columns and combine all necessary content.

In [13]:
# read df234 from the previously saved csv file
df234 = pd.read_csv('df234.csv')

# drop unnecessary column ( index from reading csv df234 )
df234 = df234.drop(['Unnamed: 0'],axis =1)

/Users/angela/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (1,3,4,5,6,7,8) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [14]:
# look at the data shape & verify whether it's the same as when we joined
df234.shape

(268706, 15)

## Create len col for each

Before we start, we wanted to do a little clean up on punctuations and malformed words. 

Then we will create length on each column to compare whether on that one row, the len in each columns are also the same.

If the length is the same, we can conclude that the values among those columns are the same, so can can drop any 2 and keep one.

If the length is not, we will look into more detail, and combine any values that are different in one row into 1 cell. 

In [15]:
# check if both product_full_name and name and brand_name have the same value in each column

# remove punctuation and malformed words in each column
df234['name']=df234['name'].str.replace(r'(\bOLD\s?\b|\n|\r|\b.\xa0)', '')
df234['name']=df234['name'].str.replace(r'\xa0',' ')

df234['brand_name']=df234['brand_name'].str.replace(r'(\bOLD\s?\b|\n|\r|\b.\xa0)', '')
df234['brand_name']=df234['brand_name'].str.replace(r'\xa0',' ')

df234['product_full_name']=df234['product_full_name'].str.replace(r'(\bOLD\s?\b|\n|\r|\b.\xa0)', '')
df234['product_full_name']=df234['product_full_name'].str.replace(r'\xa0',' ')


# create a length column for each of above column mentioned
df234['product_full_name_len'] = df234['product_full_name'].str.len()
df234['name_len'] = df234['name'].str.len()
df234['brand_name_len'] = df234['brand_name'].str.len()

# replace NaN with 0
df234['name_len'] = df234['name_len'].fillna(0)
df234['brand_name_len'] = df234['brand_name_len'].fillna(0)
df234['product_full_name_len'] = df234['product_full_name_len'].fillna(0)


# look at the difference in length
df234['len_name_brandname'] = df234['name_len']-df234['brand_name_len']


df234['len_name_brandname'].sum()

# drop column name, name_len, and len_name_brand_name
# df234= df234.drop(['name','name_len','len_name_brand_name'], axis=1)




/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  after removing the cwd from sys.path.
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: The default value of regex will change from True to False in a future version.
  import sys
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: The default value of regex will change from True to False in a future version.
  
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: FutureWarning: The default value of regex will change from True to False in a future version.
  # Remove the CWD from sys.path while we load stuff.
/Users/a

-402.0

if the sum of column len_name_brandname = 0, it means that name and name are brand_name the same

but since they are not, sum = -402, there are some rows which have different value

we will create a new col & combine those value into 1 cell

for those with the same values on both col, we will just display either one

## Create `name_2` from col `name`

First, we will look at columns `product_full_name` and `name`. Since `product_full_name` seems to contain more info in its column, we decide to take any value from `product_full_name` and fill in column `name` if any cell in `name` happen to be `NaN`


In [16]:
# fill in NaN value with product_full_name value

# define a function for 2 columns product_full_name, name

def test(product_full_name, product_full_name_len, name, name_len):
    if name_len == 0 :
        return product_full_name
    else:
        return name

# create a col name_2 that displays name value
# if name_len = zero, then display value in product_full_name, otherwise, use value in name column

df234['name_2'] = df234.apply(lambda x: test(x['product_full_name'],x['product_full_name_len'],x['name'], x['name_len'] ), axis=1)

# #check 
# df234.loc[df234['brand_name_len'] != df234['name_len']]

#drop unused col & retain columns to retain 'name_2'
df234 = df234.drop(['name','name_len'], axis=1) 

df234

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x   product_full_name  \
0                bottom  Eileen Fisher     Slim Knit Skirt   
1                bottom  Eileen Fisher     Slim Knit Skirt   
2                bottom  Eileen Fisher     Slim Knit Skirt   
3                bottom  Eileen Fisher     Slim Knit Skirt   
4                   top  Eileen Fisher  Rib Mock Neck Tank   
...                 ...            ...                 ...   
268701              NaN            NaN                 NaN   
268702              NaN            NaN                 NaN   
268703              NaN            NaN                 NaN   
268704              NaN            NaN                 NaN   
268705              NaN            NaN                 NaN   

                  product_color_id attribute_name      attribute_value  \
0       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
1       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
2       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
3       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
4       01DMBRYVA23E5A8GBES8QJN3HF       category  blazerscoatsjackets   
...                            ...            ...                  ...   
268701                         NaN            NaN                  NaN   
268702                         NaN            NaN                  NaN   
268703                         NaN            NaN                  NaN   
268704                         NaN            NaN                  NaN   
268705                         NaN            NaN                  NaN   

        brand_y        brand_category  \
0           NaN                   NaN   
1           NaN                   NaN   
2           NaN                   NaN   
3           NaN                   NaN   
4           NaN                   NaN   
...         ...                   ...   
268701  Nomasei          SandalsSales   
268702  Nomasei  Flat sandalsArchives   
268703  Nomasei  Flat sandalsArchives   
268704  Nomasei          SandalsSales   
268705  Nomasei  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...   

## Create `brand_name_2` from col `brand_name`

Similar to `name_2`, we also create new column `brand_name_2` from `brand_name` with the same context

If the length in `brand_name` is zero, we will fill in the value from `product_full_name`. Otherwise, we will return the exact same value of `brand_name`

In [17]:
# define a function for 2 columns product_full_name,brand_name

def test(product_full_name, product_full_name_len, brand_name, brand_name_len):
    if brand_name_len == 0 :
        return product_full_name
    else:
        return brand_name
# create a col brandname_name_combine that displays brand_name/name value
# if brand_name and name col have different value, then display both in 1 cell

df234['brand_name_2'] = df234.apply(lambda x: test(x['product_full_name'],x['product_full_name_len'],x['brand_name'], x['brand_name_len'] ), axis=1)

# #check
# df234.loc[df234['brand_name_len'] != df234['name_len']]

# #drop unused col & retain columns to retain 'name_2'
# df234 = df234.drop(['brand_name','brand_name_len','product_full_name_len'], axis=1)

df234

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x   product_full_name  \
0                bottom  Eileen Fisher     Slim Knit Skirt   
1                bottom  Eileen Fisher     Slim Knit Skirt   
2                bottom  Eileen Fisher     Slim Knit Skirt   
3                bottom  Eileen Fisher     Slim Knit Skirt   
4                   top  Eileen Fisher  Rib Mock Neck Tank   
...                 ...            ...                 ...   
268701              NaN            NaN                 NaN   
268702              NaN            NaN                 NaN   
268703              NaN            NaN                 NaN   
268704              NaN            NaN                 NaN   
268705              NaN            NaN                 NaN   

                  product_color_id attribute_name      attribute_value  \
0       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
1       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
2       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
3       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
4       01DMBRYVA23E5A8GBES8QJN3HF       category  blazerscoatsjackets   
...                            ...            ...                  ...   
268701                         NaN            NaN                  NaN   
268702                         NaN            NaN                  NaN   
268703                         NaN            NaN                  NaN   
268704                         NaN            NaN                  NaN   
268705                         NaN            NaN                  NaN   

        brand_y        brand_category  \
0           NaN                   NaN   
1           NaN                   NaN   
2           NaN                   NaN   
3           NaN                   NaN   
4           NaN                   NaN   
...         ...                   ...   
268701  Nomasei          SandalsSales   
268702  Nomasei  Flat sandalsArchives   
268703  Nomasei  Flat sandalsArchives   
268704  Nomasei          SandalsSales   
268705  Nomasei  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...   

## Verify if `name_2` and `brand_name_2` are exactly the same 

Now, we look into `name_2` and `brand_name_2` and compare the length whether all rows both have the same length.
 
If they are, then we can conclude that `name_2` and `brand_name_2` are the same, and we con drop either one

If they are not, we will look into detail, and combine any rows with different value

In [18]:
# verify if name_2 and brand_name_2 are exactly the same 

df234['name_len'] = df234['name_2'].str.len()
df234['brand_name_len'] = df234['brand_name_2'].str.len()

# replace NaN with 0
df234['name_len'] = df234['name_len'].fillna(0)
df234['brand_name_len'] = df234['brand_name_len'].fillna(0)


# look at the difference in length
df234['len_name_brandname'] = df234['name_len']-df234['brand_name_len']
df234['len_name_brandname'].sum()

-402.0

since they are not, sum = -402, there are some rows which have different value

we will create a new column called `brandname_name_combine` & combine those value into 1 cell

for those with the same values on both col, we will just display either one

In [19]:
# looking at the difference between brand_name_len and name_len
df234[df234['brand_name_len']!=df234['name_len']]

outfit_id                  product_id outfit_item_type brand_x  \
234487       NaN  01F1QF61C43NBYGNWC162VEQ63              NaN     NaN   
234488       NaN  01F1QF65XE634V0A31MFFW62WT              NaN     NaN   
234489       NaN  01F1QF61F1Y9CJXGR9F6JYX0F3              NaN     NaN   
234507       NaN  01F1QF6EHD3J05BE3R0D1A1TSP              NaN     NaN   
234517       NaN  01F1QF66A5FZ79WNM3S86SFCGF              NaN     NaN   
...          ...                         ...              ...     ...   
261893       NaN  01ESS0TNGGD6JWDV2X5JH7G739              NaN     NaN   
261894       NaN  01ESS0THVHG003798PV9HRB33H              NaN     NaN   
264448       NaN  01ESS0NTG5AE5DDCQZMJ23014D              NaN     NaN   
264449       NaN  01ESS0REAB07Z0P1WAJP1RG61H              NaN     NaN   
264608       NaN  01EY3GV2D4DHB1A2NZWAN44JSW              NaN     NaN   

       product_full_name product_color_id attribute_name attribute_value  \
234487               NaN              NaN            NaN             NaN   
234488               NaN              NaN            NaN             NaN   
234489               NaN              NaN            NaN             NaN   
234507               NaN              NaN            NaN             NaN   
234517               NaN              NaN            NaN             NaN   
...                  ...              ...            ...             ...   
261893               NaN              NaN            NaN             NaN   
261894               NaN              NaN            NaN             NaN   
264448               NaN              NaN            NaN             NaN   
264449               NaN              NaN            NaN             NaN   
264608               NaN              NaN            NaN             NaN   

                brand_y brand_category details  \
234487  Mari Giudicelli        Unknown     NaN   
234488  Mari Giudicelli        Unknown     NaN   
234489  Mari Giudicelli        Unknown     NaN   
234507  Mari Giudicelli        Unknown     NaN   
234517  Mari Giudicelli        Unknown     NaN   
...                 ...            ...     ...   
261893  Mari Giudicelli        Unknown     NaN   
261894  Mari Giudicelli        Unknown     NaN   
264448  Mari Giudicelli        Unknown     NaN   
264449  Mari Giudicelli        Unknown     NaN   
264608  Mari Giudicelli        Unknown     NaN   

                                              description  \
234487  Black Sheepskin Leather. Wraparound ankle stra...   
234488  Black Sheepskin Braided Leather.\nLeather Lini...   
234489  Cereja Sheepskin Braided Leather.\nLeather Lin...   
234507  This style runs slightly big.\nVerde Sheepskin...   
234517  Natural Cotton Open Toe Sandal.\nAdjustable An...   
...                                                   ...   
261893  This style runs true to size.\nDeep Red Woven ...   
261894  This style runs true to size.\nDeep Red Lambsk...   
264448  Off White Woven Sheepskin Leather. Pointed Toe...   
264449  This style runs slightly big.\nOff White Sheep...   
264608  This style runs true to size.\nBlack Sheepskin...   

                                        brand_description  \
234487  Black Sheepskin Leather. Wraparound ankle stra...   
234488  Black Sheepskin Braided Leather.\nLeather Lini...   
234489  Cereja Sheepskin Braided Leather.\nLeather Lin...   
234507  This style runs slightly big.\n\nVerde Sheepsk...   
234517  Natural Cotton Open Toe Sandal.\nAdjustable An...   
...                                                   ...   
261893  This style runs true to size.\nDeep Red Woven ...   
261894  This style runs true to size.\n\nDeep Red Lamb...   
264448  Off White Woven Sheepskin Leather. Pointed Toe...   
264449  This style runs slightly big.\n\nOff White She...   
264608  This style runs true to size.\n\nBlack Sheepsk...   

                       brand_name  product_full_name_len  brand_name_len  \
234487            Black Sheepskin                    0.0            15.0  

In [20]:
# define a function for 2 columns name_2,brand_name_2

def test(name_2, brand_name_2, name_len, brand_name_len):
    if brand_name_len == name_len:
        return brand_name_2
    else:
        return name_2, brand_name_2
# create a col brandname_name_combine that displays brand_name/name value
# if brand_name and name col have different value, then display both in 1 cell

df234['brandname_name_combine'] = df234.apply(lambda x: test(x['name_2'],x['brand_name_2'],x['name_len'], x['brand_name_len'] ), axis=1)
# check
# df234.loc[df234['brand_name_len'] != df234['name_len']]
# df234 = df234.drop(['brand_name','brand_name_len','product_full_name_len'], axis=1)

df234

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x   product_full_name  \
0                bottom  Eileen Fisher     Slim Knit Skirt   
1                bottom  Eileen Fisher     Slim Knit Skirt   
2                bottom  Eileen Fisher     Slim Knit Skirt   
3                bottom  Eileen Fisher     Slim Knit Skirt   
4                   top  Eileen Fisher  Rib Mock Neck Tank   
...                 ...            ...                 ...   
268701              NaN            NaN                 NaN   
268702              NaN            NaN                 NaN   
268703              NaN            NaN                 NaN   
268704              NaN            NaN                 NaN   
268705              NaN            NaN                 NaN   

                  product_color_id attribute_name      attribute_value  \
0       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
1       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
2       01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor               beiges   
3       01DMBRYVA2C7BQN0X7MNNYXBHW       category               bottom   
4       01DMBRYVA23E5A8GBES8QJN3HF       category  blazerscoatsjackets   
...                            ...            ...                  ...   
268701                         NaN            NaN                  NaN   
268702                         NaN            NaN                  NaN   
268703                         NaN            NaN                  NaN   
268704                         NaN            NaN                  NaN   
268705                         NaN            NaN                  NaN   

        brand_y        brand_category  \
0           NaN                   NaN   
1           NaN                   NaN   
2           NaN                   NaN   
3           NaN                   NaN   
4           NaN                   NaN   
...         ...                   ...   
268701  Nomasei          SandalsSales   
268702  Nomasei  Flat sandalsArchives   
268703  Nomasei  Flat sandalsArchives   
268704  Nomasei          SandalsSales   
268705  Nomasei  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...   

In [21]:
# df234['len_x_y'] = df234['product_full_name_len']-df234['name_len']

# we look at the different in `brand_name_2` and `name_2`
df234[df234['brand_name_len']!=df234['name_len']]

outfit_id                  product_id outfit_item_type brand_x  \
234487       NaN  01F1QF61C43NBYGNWC162VEQ63              NaN     NaN   
234488       NaN  01F1QF65XE634V0A31MFFW62WT              NaN     NaN   
234489       NaN  01F1QF61F1Y9CJXGR9F6JYX0F3              NaN     NaN   
234507       NaN  01F1QF6EHD3J05BE3R0D1A1TSP              NaN     NaN   
234517       NaN  01F1QF66A5FZ79WNM3S86SFCGF              NaN     NaN   
...          ...                         ...              ...     ...   
261893       NaN  01ESS0TNGGD6JWDV2X5JH7G739              NaN     NaN   
261894       NaN  01ESS0THVHG003798PV9HRB33H              NaN     NaN   
264448       NaN  01ESS0NTG5AE5DDCQZMJ23014D              NaN     NaN   
264449       NaN  01ESS0REAB07Z0P1WAJP1RG61H              NaN     NaN   
264608       NaN  01EY3GV2D4DHB1A2NZWAN44JSW              NaN     NaN   

       product_full_name product_color_id attribute_name attribute_value  \
234487               NaN              NaN            NaN             NaN   
234488               NaN              NaN            NaN             NaN   
234489               NaN              NaN            NaN             NaN   
234507               NaN              NaN            NaN             NaN   
234517               NaN              NaN            NaN             NaN   
...                  ...              ...            ...             ...   
261893               NaN              NaN            NaN             NaN   
261894               NaN              NaN            NaN             NaN   
264448               NaN              NaN            NaN             NaN   
264449               NaN              NaN            NaN             NaN   
264608               NaN              NaN            NaN             NaN   

                brand_y brand_category details  \
234487  Mari Giudicelli        Unknown     NaN   
234488  Mari Giudicelli        Unknown     NaN   
234489  Mari Giudicelli        Unknown     NaN   
234507  Mari Giudicelli        Unknown     NaN   
234517  Mari Giudicelli        Unknown     NaN   
...                 ...            ...     ...   
261893  Mari Giudicelli        Unknown     NaN   
261894  Mari Giudicelli        Unknown     NaN   
264448  Mari Giudicelli        Unknown     NaN   
264449  Mari Giudicelli        Unknown     NaN   
264608  Mari Giudicelli        Unknown     NaN   

                                              description  \
234487  Black Sheepskin Leather. Wraparound ankle stra...   
234488  Black Sheepskin Braided Leather.\nLeather Lini...   
234489  Cereja Sheepskin Braided Leather.\nLeather Lin...   
234507  This style runs slightly big.\nVerde Sheepskin...   
234517  Natural Cotton Open Toe Sandal.\nAdjustable An...   
...                                                   ...   
261893  This style runs true to size.\nDeep Red Woven ...   
261894  This style runs true to size.\nDeep Red Lambsk...   
264448  Off White Woven Sheepskin Leather. Pointed Toe...   
264449  This style runs slightly big.\nOff White Sheep...   
264608  This style runs true to size.\nBlack Sheepskin...   

                                        brand_description  \
234487  Black Sheepskin Leather. Wraparound ankle stra...   
234488  Black Sheepskin Braided Leather.\nLeather Lini...   
234489  Cereja Sheepskin Braided Leather.\nLeather Lin...   
234507  This style runs slightly big.\n\nVerde Sheepsk...   
234517  Natural Cotton Open Toe Sandal.\nAdjustable An...   
...                                                   ...   
261893  This style runs true to size.\nDeep Red Woven ...   
261894  This style runs true to size.\n\nDeep Red Lamb...   
264448  Off White Woven Sheepskin Leather. Pointed Toe...   
264449  This style runs slightly big.\n\nOff White She...   
264608  This style runs true to size.\n\nBlack Sheepsk...   

                       brand_name  product_full_name_len  brand_name_len  \
234487            Black Sheepskin                    0.0            15.0  

## Combine 3 columns into `brandname_name_combine`

In [22]:
# checking NaN value
df234['brandname_name_combine'].isna().sum()

3171

In [23]:
# create len for brandname_name_combine col
df234['brandname_name_combine_len'] = df234['brandname_name_combine'].str.len()

In [24]:
df234['product_full_name_len'].fillna(0)
# define a function for 3 columns product_full_name, brandname_name_combine

def test(product_full_name, brandname_name_combine, product_full_name_len, brandname_name_combine_len):
    if product_full_name_len == brandname_name_combine_len:
        return brandname_name_combine
    elif product_full_name_len == 0:
        return brandname_name_combine
    else:
        return product_full_name, brandname_name_combine
# create a col brandname_name_combine that displays brand_name/name value
# if brand_name and name col have different value, then display both in 1 cell
# if they have the same value, we will only display either one value (not twice)

df234['product_brand_name_combine'] = df234.apply(lambda x: test(x['product_full_name'],x['brandname_name_combine'], x['product_full_name_len'],x['brandname_name_combine_len'] ), axis=1)
# check
df234.loc[df234['product_full_name_len'] != df234['brandname_name_combine_len']]
# df234 = df234.drop(['brand_name','brand_name_len','product_full_name_len'], axis=1)

# df234

outfit_id                  product_id  \
43322   01DVBSVM1E0WZD41A07K4K0T14  01DT0DK5JZGAYR0C8983NCA7T5   
43323   01DVBSVM1E0WZD41A07K4K0T14  01DT0DK5JZGAYR0C8983NCA7T5   
43324   01DVBSVM1E0WZD41A07K4K0T14  01DT0DK5JZGAYR0C8983NCA7T5   
43325   01DVBSVM1E0WZD41A07K4K0T14  01DT0DK5JZGAYR0C8983NCA7T5   
43326   01DVBSVM1E0WZD41A07K4K0T14  01DT0DK5JZGAYR0C8983NCA7T5   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x  product_full_name  \
43322        accessory1  Gigi New York  Tori Leather Tote   
43323        accessory1  Gigi New York  Tori Leather Tote   
43324        accessory1  Gigi New York  Tori Leather Tote   
43325        accessory1  Gigi New York  Tori Leather Tote   
43326        accessory1  Gigi New York  Tori Leather Tote   
...                 ...            ...                ...   
268701              NaN            NaN                NaN   
268702              NaN            NaN                NaN   
268703              NaN            NaN                NaN   
268704              NaN            NaN                NaN   
268705              NaN            NaN                NaN   

                  product_color_id attribute_name attribute_value  \
43322   01DT0DKDNX1P8KK0GA55HVSGA3   primarycolor           blues   
43323   01DT0DKDNX1P8KK0GA55HVSGA3        pattern          animal   
43324   01DT0DKDNX1P8KK0GA55HVSGA3          style     androgynous   
43325   01DT0DKDNX1P8KK0GA55HVSGA3  strapmaterial         leather   
43326   01DT0DKDNX1P8KK0GA55HVSGA3          style         classic   
...                            ...            ...             ...   
268701                         NaN            NaN             NaN   
268702                         NaN            NaN             NaN   
268703                         NaN            NaN             NaN   
268704                         NaN            NaN             NaN   
268705                         NaN            NaN             NaN   

              brand_y           brand_category  \
43322   Gigi New York  Handbags/Handbags/Totes   
43323   Gigi New York  Handbags/Handbags/Totes   
43324   Gigi New York  Handbags/Handbags/Totes   
43325   Gigi New York  Handbags/Handbags/Totes   
43326   Gigi New York  Handbags/Handbags/Totes   
...               ...                      ...   
268701        Nomasei             SandalsSales   
268702        Nomasei     Flat sandalsArchives   
268703        Nomasei     Flat sandalsArchives   
268704        Nomasei             SandalsSales   
268705        Nomasei     Flat sandalsArchives   

                                                  details  \
43322                                                 NaN   
43323                                                 NaN   
43324                                                 NaN   
43325                                                 NaN   
43326                                                 NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
43322   Soft leather tote opens to an unlined interior...   
43323   Soft leather tote opens to an unlined interior...   
43324   Soft leather tote opens to an unlined interior...   
43325   Soft leather tote opens to an unlined interior...   
43326   Soft leather tote opens

In [26]:
# remove those columns (unnecessary), only keep product_brand_name_combine col
df234 = df234.drop(['product_full_name','brand_name','product_full_name_len','brand_name_len','name_len','name_2','brand_name_2','len_name_brandname','brandname_name_combine','brandname_name_combine_len'], axis=1)

# Combine `brand_x` vs `brand_y` into `brand`

## checking duplicate values / diff values

We will the same method and context as above to combine those 2 columns into 1

In [27]:
# verify if name_2 and brand_name_2 are exactly the same 

df234['brand_x_len'] = df234['brand_x'].str.len()
df234['brand_y_len'] = df234['brand_y'].str.len()

# replace NaN with 0
df234['brand_x_len'] = df234['brand_x_len'].fillna(0)
df234['brand_y_len'] = df234['brand_y_len'].fillna(0)

# look at the difference in length
df234['len_x_y'] = df234['brand_x_len']-df234['brand_y_len']
df234['len_x_y'].sum()

-1241721.0

In [28]:
# checking the difference between brand_x and brand_y
df234[df234['brand_x_len']!=df234['brand_y_len']]

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x            product_color_id  \
0                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
1                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
2                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
3                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
4                   top  Eileen Fisher  01DMBRYVA23E5A8GBES8QJN3HF   
...                 ...            ...                         ...   
268701              NaN            NaN                         NaN   
268702              NaN            NaN                         NaN   
268703              NaN            NaN                         NaN   
268704              NaN            NaN                         NaN   
268705              NaN            NaN                         NaN   

       attribute_name      attribute_value  brand_y        brand_category  \
0        primarycolor               beiges      NaN                   NaN   
1            category               bottom      NaN                   NaN   
2        primarycolor               beiges      NaN                   NaN   
3            category               bottom      NaN                   NaN   
4            category  blazerscoatsjackets      NaN                   NaN   
...               ...                  ...      ...                   ...   
268701            NaN                  NaN  Nomasei          SandalsSales   
268702            NaN                  NaN  Nomasei  Flat sandalsArchives   
268703            NaN                  NaN  Nomasei  Flat sandalsArchives   
268704            NaN                  NaN  Nomasei          SandalsSales   
268705            NaN                  NaN  Nomasei  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  The flat BowVida mule in black suede is the id...   
268702  The flat Vida mule in tangerine suede is the c...   
268703  The flat BowVida mule in fuschia suede is the ...   
268704  The flat Vida mule in silver metalized leather...   
268705  The flat BowVida mule in bronze leather is the...   



## combine col `brand_x` and `brand_y` into `brand`

In [29]:
# define a function for 2 columns brand_x, brand_y

def test(brand_x, brand_y, brand_x_len, brand_y_len):
    if brand_x_len == brand_y_len:
        return brand_y
    elif brand_x_len == 0: # if brand_x is NaN, then display the value of brand_y
        return brand_y
    elif brand_y_len == 0: # if brand_y is NaN, then display the value of brand_x
        return brand_x
    else:
        return brand_x, brand_y # if brand_x and brand_y col have different value, then display both in 1 cell

# create a col brand that displays brand_x/brand_y
df234['brand'] = df234.apply(lambda x: test(x['brand_x'],x['brand_y'], x['brand_x_len'],x['brand_y_len'] ), axis=1)
# check the difference between brand_x and brand_y
df234.loc[df234['brand_x_len'] != df234['brand_y_len']]

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type        brand_x            product_color_id  \
0                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
1                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
2                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
3                bottom  Eileen Fisher  01DMBRYVA2C7BQN0X7MNNYXBHW   
4                   top  Eileen Fisher  01DMBRYVA23E5A8GBES8QJN3HF   
...                 ...            ...                         ...   
268701              NaN            NaN                         NaN   
268702              NaN            NaN                         NaN   
268703              NaN            NaN                         NaN   
268704              NaN            NaN                         NaN   
268705              NaN            NaN                         NaN   

       attribute_name      attribute_value  brand_y        brand_category  \
0        primarycolor               beiges      NaN                   NaN   
1            category               bottom      NaN                   NaN   
2        primarycolor               beiges      NaN                   NaN   
3            category               bottom      NaN                   NaN   
4            category  blazerscoatsjackets      NaN                   NaN   
...               ...                  ...      ...                   ...   
268701            NaN                  NaN  Nomasei          SandalsSales   
268702            NaN                  NaN  Nomasei  Flat sandalsArchives   
268703            NaN                  NaN  Nomasei  Flat sandalsArchives   
268704            NaN                  NaN  Nomasei          SandalsSales   
268705            NaN                  NaN  Nomasei  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  The flat BowVida mule in black suede is the id...   
268702  The flat Vida mule in tangerine suede is the c...   
268703  The flat BowVida mule in fuschia suede is the ...   
268704  The flat Vida mule in silver metalized leather...   
268705  The flat BowVida mule in bronze leather is the...   



In [30]:
# remove unnecessary col
df234 = df234.drop(['brand_x','brand_x_len','brand_y','brand_y_len'], axis=1)
df234

outfit_id                  product_id  \
0       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
1       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2P5H24WK0HTK4R0A1   
2       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
3       01DMHCX50CFX5YNG99F3Y65GQW  01DMBRYVA2P5H24WK0HTK4R0A1   
4       01DDBHC62ES5K80P0KYJ56AM2T  01DMBRYVA2PEPWFTT7RMP5AA1T   
...                            ...                         ...   
268701                         NaN  01EYB5ERGYPFNGM6C9QK7Q9EV0   
268702                         NaN  01EHWTBFP368Q035FW95TRJDAY   
268703                         NaN  01EHWTCFTPPSCW10D4XBQZF28H   
268704                         NaN  01EYB5B5FH7JESXF82ZEMVXZMS   
268705                         NaN  01EHWTF4VNZ6FPF2836A0CFEEG   

       outfit_item_type            product_color_id attribute_name  \
0                bottom  01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor   
1                bottom  01DMBRYVA2C7BQN0X7MNNYXBHW       category   
2                bottom  01DMBRYVA2C7BQN0X7MNNYXBHW   primarycolor   
3                bottom  01DMBRYVA2C7BQN0X7MNNYXBHW       category   
4                   top  01DMBRYVA23E5A8GBES8QJN3HF       category   
...                 ...                         ...            ...   
268701              NaN                         NaN            NaN   
268702              NaN                         NaN            NaN   
268703              NaN                         NaN            NaN   
268704              NaN                         NaN            NaN   
268705              NaN                         NaN            NaN   

            attribute_value        brand_category  \
0                    beiges                   NaN   
1                    bottom                   NaN   
2                    beiges                   NaN   
3                    bottom                   NaN   
4       blazerscoatsjackets                   NaN   
...                     ...                   ...   
268701                  NaN          SandalsSales   
268702                  NaN  Flat sandalsArchives   
268703                  NaN  Flat sandalsArchives   
268704                  NaN          SandalsSales   
268705                  NaN  Flat sandalsArchives   

                                                  details  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  Feminine flat mules.\nSquare shape, V line on ...   
268702  Feminine flat mules.\nSquare shape, V line on ...   
268703  Feminine flat mules.\nSquare shape, V line on ...   
268704  Feminine flat mules.\nSquare shape, V line on ...   
268705  Feminine flat mules.\nSquare shape, V line on ...   

                                              description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
268701  The flat BowVida mule in black suede is the id...   
268702  The flat Vida mule in tangerine suede is the c...   
268703  The flat BowVida mule in fuschia suede is the ...   
268704  The flat Vida mule in silver metalized leather...   
268705  The flat BowVida mule in bronze leather is the...   

                                        brand_description  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                           

In [31]:
df234.to_csv('df234.csv')

# Combine `brand_description` and `description` into `branddescription_description_combine`

We will the same method and context as above to combine those 2 columns into 1

## checking duplicate col `description` vs `brand_description` in file 2

In [32]:
# read csv file
df234=pd.read_csv('df234.csv')

/Users/angela/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (1,3,4,5,6) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [33]:
# remove stopwords/punctuation/ malformed words 
df234['description']=df234['description'].str.replace(r'(\bOLD\s?\b|\n|\r|\*|\b.\xa0|[^a-z|^0-9](.\s?$)|[^a-z|^0-9]($))', '')
df234['description']=df234['description'].str.replace(r'\xa0',' ')
df234['brand_description']=df234['brand_description'].str.replace(r'(\bOLD\s?\b|\n|\r|\*|\b.\xa0|[^a-z|^0-9](.\s?$)|[^a-z|^0-9]($))', '')
df234['brand_description']=df234['brand_description'].str.replace(r'\xa0',' ')


# check if both description and brand_description have the same value in each column
df234['description_len'] = df234['description'].str.len()
df234['brand_description_len'] = df234['brand_description'].str.len()

# replace NaN with zero
df234['description_len'] = df234['description_len'].fillna(0)
df234['brand_description_len'] = df234['brand_description_len'].fillna(0)

# look at the difference in length
df234['len_x_y'] = df234['description_len']-df234['brand_description_len']
# if the sum of column len_x_y = 0, it means brand_x and brand_y are exactly the same, we okay to drop either one

df234['len_x_y'].sum()


/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  after removing the cwd from sys.path.
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """


-747242.0

In [34]:
df234.loc[df234['brand_description_len']!=df234['description_len']]
# some rows have a huge difference eg. 238632, 222020 etc.
# total of 34243 rows with huge diff

Unnamed: 0                   outfit_id                  product_id  \
21480        21480  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21481        21481  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21482        21482  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21483        21483  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21484        21484  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
...            ...                         ...                         ...   
268539      268539                         NaN  01EB2DZA7QGHNT0GPCKJVBAZZD   
268560      268560                         NaN  01EB2DSJK7F24RCSDTNT4J2Z16   
268582      268582                         NaN  01EB2DNEEPH2H7KW7GMWR675DN   
268602      268602                         NaN  01EB2D5WNFARVGJ6SDR3QK39MY   
268630      268630                         NaN  01EB2DGQ9C9R3TWXJZJX2C6B9D   

       outfit_item_type            product_color_id    attribute_name  \
21480               top  01DTJ8G50ZHR71CCA3E6QAAH5G      primarycolor   
21481               top  01DTJ8G50ZHR71CCA3E6QAAH5G             style   
21482               top  01DTJ8G50ZHR71CCA3E6QAAH5G             style   
21483               top  01DTJ8G50ZHR71CCA3E6QAAH5G  materialclothing   
21484               top  01DTJ8G50ZHR71CCA3E6QAAH5G  materialclothing   
...                 ...                         ...               ...   
268539              NaN                         NaN               NaN   
268560              NaN                         NaN               NaN   
268582              NaN                         NaN               NaN   
268602              NaN                         NaN               NaN   
268630              NaN                         NaN               NaN   

       attribute_value brand_category details  \
21480           greens      Jumpsuits     NaN   
21481           casual      Jumpsuits     NaN   
21482           modern      Jumpsuits     NaN   
21483         elastane      Jumpsuits     NaN   
21484      cottonblend      Jumpsuits     NaN   
...                ...            ...     ...   
268539             NaN      Jumpsuits     NaN   
268560             NaN           Tops     NaN   
268582             NaN         Shorts     NaN   
268602             NaN          Pants     NaN   
268630             NaN          Pants     NaN   

                                              description  \
21480   Cut from army-green utility twill, Julian is o...   
21481   Cut from army-green utility twill, Julian is o...   
21482   Cut from army-green utility twill, Julian is o...   
21483   Cut from army-green utility twill, Julian is o...   
21484   Cut from army-green utility twill, Julian is o...   
...                                                   ...   
268539  Fit:Waisted SilhouetteEasy Wide Leg27" Ankle S...   
268560  Fit:Easy fitting crewneck bodysuitFabric:100% ...   
268582  Fit:Mid Rise Relaxed Short Easy fit through th...   
268602  Fit:High Rise  Relaxed Trouser Vintage Button ...   
268630  Fit: Mid Rise Slim BootcutSubtly Flare below t...   

                                        brand_description  \
21480   Cut from army-green utility twill, Julian is o...   
21481   Cut from army-green utility twill, Julian is o...   
21482   Cut from army-green utility twill, Julian is o...   
21483   Cut from army-green utility twill, Julian is o...   
21484   Cut from army-green utility twill, Julian is o...   
...                                                   ...   
268539  Fit:Waisted SilhouetteEasy Wide Leg27" Ankle S...   
268560  Fit:Easy fitting crewneck bodysuitFabric:100Su...   
268582  Fit:Mid Rise Relaxed Short Easy fit through th...   
268602  Fit:High Rise  Relaxed Trouser Vintage Button ...   
268630  FitMid Rise Slim BootcutSubtly Flare below the...   

         product_brand_name_combine  len_x_y                 brand  \
21480      Julian Workwear Jumpsuit      2.0              G. Label   

In [35]:
# # testing regex sub
# import re
# test = 'Consider tucking yours into a pair of high-waisted trousers to create a more defined shape. '
# re.sub(r'[^a-z|^0-9](.\s?$|$)', '',test)
# # test = test.replace('^a-z|^0-9](.\s?$)','')

## combine `brand_description` and `description`

We will create a new column called `branddescription_description_combine` which contain the value of both brand_description and description with below conditions

If the lengths are the same, in each row, we can display either value from brand_description/ description

If the length are different, we will combine those into 1 value like what we did earlier for `product_full_name` and `name_2`

However, if the length difference is less than 5, we can still assume those are basically the same context since there may be there different in typing/ typo/ plurals etc. those shouldn't be a huge effect on the  meanning of the whole sentence


In [36]:
# define a function for 2 columns brand_description,description

def test(brand_description, brand_description_len, description, description_len, len_x_y):
    if brand_description_len == description_len :
        return brand_description # if brand_name and name col have same value, display either one (in this case, use brand_description value)
    elif abs(len_x_y) <= 5:
        return brand_description # if the difference in length is less than 5 characters, we can still take it as the same value and display either one
    else:
        return brand_description, description # if brand_name and name col have different value, then display both in 1 cell

# create a col brandname_name_combine that displays brand_name/name value
df234['branddescription_description_combine'] = df234.apply(lambda x: test(x['brand_description'], x['brand_description_len'],x['description'], x['description_len'], x['len_x_y'] ), axis=1)
# check
df234.loc[df234['brand_description_len'] != df234['description_len']]


Unnamed: 0                   outfit_id                  product_id  \
21480        21480  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21481        21481  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21482        21482  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21483        21483  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
21484        21484  01DTMXEHJJ0Z7W53VVZ7SBP50X  01DTJ8G50MTEAZ269BY9PDK0MC   
...            ...                         ...                         ...   
268539      268539                         NaN  01EB2DZA7QGHNT0GPCKJVBAZZD   
268560      268560                         NaN  01EB2DSJK7F24RCSDTNT4J2Z16   
268582      268582                         NaN  01EB2DNEEPH2H7KW7GMWR675DN   
268602      268602                         NaN  01EB2D5WNFARVGJ6SDR3QK39MY   
268630      268630                         NaN  01EB2DGQ9C9R3TWXJZJX2C6B9D   

       outfit_item_type            product_color_id    attribute_name  \
21480               top  01DTJ8G50ZHR71CCA3E6QAAH5G      primarycolor   
21481               top  01DTJ8G50ZHR71CCA3E6QAAH5G             style   
21482               top  01DTJ8G50ZHR71CCA3E6QAAH5G             style   
21483               top  01DTJ8G50ZHR71CCA3E6QAAH5G  materialclothing   
21484               top  01DTJ8G50ZHR71CCA3E6QAAH5G  materialclothing   
...                 ...                         ...               ...   
268539              NaN                         NaN               NaN   
268560              NaN                         NaN               NaN   
268582              NaN                         NaN               NaN   
268602              NaN                         NaN               NaN   
268630              NaN                         NaN               NaN   

       attribute_value brand_category details  \
21480           greens      Jumpsuits     NaN   
21481           casual      Jumpsuits     NaN   
21482           modern      Jumpsuits     NaN   
21483         elastane      Jumpsuits     NaN   
21484      cottonblend      Jumpsuits     NaN   
...                ...            ...     ...   
268539             NaN      Jumpsuits     NaN   
268560             NaN           Tops     NaN   
268582             NaN         Shorts     NaN   
268602             NaN          Pants     NaN   
268630             NaN          Pants     NaN   

                                              description  \
21480   Cut from army-green utility twill, Julian is o...   
21481   Cut from army-green utility twill, Julian is o...   
21482   Cut from army-green utility twill, Julian is o...   
21483   Cut from army-green utility twill, Julian is o...   
21484   Cut from army-green utility twill, Julian is o...   
...                                                   ...   
268539  Fit:Waisted SilhouetteEasy Wide Leg27" Ankle S...   
268560  Fit:Easy fitting crewneck bodysuitFabric:100% ...   
268582  Fit:Mid Rise Relaxed Short Easy fit through th...   
268602  Fit:High Rise  Relaxed Trouser Vintage Button ...   
268630  Fit: Mid Rise Slim BootcutSubtly Flare below t...   

                                        brand_description  \
21480   Cut from army-green utility twill, Julian is o...   
21481   Cut from army-green utility twill, Julian is o...   
21482   Cut from army-green utility twill, Julian is o...   
21483   Cut from army-green utility twill, Julian is o...   
21484   Cut from army-green utility twill, Julian is o...   
...                                                   ...   
268539  Fit:Waisted SilhouetteEasy Wide Leg27" Ankle S...   
268560  Fit:Easy fitting crewneck bodysuitFabric:100Su...   
268582  Fit:Mid Rise Relaxed Short Easy fit through th...   
268602  Fit:High Rise  Relaxed Trouser Vintage Button ...   
268630  FitMid Rise Slim BootcutSubtly Flare below the...   

         product_brand_name_combine  len_x_y                 brand  \
21480      Julian Workwear Jumpsuit      2.0              G. Label   

In [37]:
# drop column brand_description_len, description_len, brand_description, description, len_x_y
df234= df234.drop(['brand_description_len', 'description_len', 'brand_description', 'description', 'len_x_y'], axis=1)

## EDA for Custom Features

Now, after we have all combine columns (as our not-yet-to-be-cleaned-up features)

We will perform Exploratory Data Analysis with an emphasis on finding unique features to help identify the different product types and brands.

### Clean up `branddescription_description_combine` and `product_brand_name_combine` columns 

In [38]:
# Initial preprocessing of target columns to perform EDA on
  # Fill in NaN values, convert all strings to lower case, and delete fully nulled out rows from data set.



# fill in NaN values
df234.fillna("", inplace=True)

# Drop any observations where every column in that row is null
df234.isnull().all(1).sum()  ## there are None


# convert strings to lowercase
df234['branddescription_description_combine'] = df234['branddescription_description_combine'].str.lower()

df234['product_brand_name_combine'] = df234['product_brand_name_combine'].str.lower()

# remove all punctuations
df234['branddescription_description_combine'] = df234['branddescription_description_combine'].str.replace(r'([^\w\d\s])', "", regex=True)

df234['product_brand_name_combine'] = df234['product_brand_name_combine'].str.replace(r'([^\w\d\s])', "", regex=True)



### Extract custom features from those above columns

Search and extract various attribute features from the combined description column. Use these features for the overall project objective



In [58]:
## fillna in the entire dataset
df234.fillna("",inplace=True)


## Color
df234['description_color']=df234['branddescription_description_combine'].str.extract(r'(suede|beige|black|blue|brown|burgundy|gold|gray|grey|\
                        green|multi|navy|neutral|orange|pinks|purple|red|silver|teal|white|yellow)')

## Attribute color
df234['attribute_color']=df234['attribute_value'].str.extract(r'(suede|beige|black|blue|brown|burgundy|gold|gray|grey|\
                        green|multi|navy|neutral|orange|pinks|purple|red|silver|teal|white|yellow)')

## Origin of clothing (Country and select fashion capital cities)
df234['country2']=df234['branddescription_description_combine'].str.extract(r'(china|italy|spain|canada|france|india|turkey|germany|japan|ghana|mali|\
peru|argentina|portugal|kenya|united states|colombia|denmark|ethiopa|philippines|madagascar|vietnam|brazil|korea|sweden|bolivia|\
nepal|mexico|usa|los.?angeles|new.?york|tokyo|\bnyc?\b)')

## Fit of the items
df234['fit1'] = df234['branddescription_description_combine'].str.extract(r'(tight?\s?.?\w{1,3}|loose?\s?.?\w{1,3}|\
loosely?\s?.?\w{1,3}|tightly?\s?.?\w{1,3}|regular?\s?.?\w{1,3}|narrow?\s?.?\w{1,3}|wide?\s?.?\w{1,3}|\
extra?\s?.?\wide?|straight?\s?.?\w{1,3}|slim\s?.?\w{1,3}|fitted?)')

## Size of the items
df234['size1'] = df234['branddescription_description_combine'].str.extract(r'\b(small|medium|large|extra\s?small|extra\s?large|xs|s|m|l|xl|xxl|xxxL|one\s?size|uni\s?size)\b')

## Body part emphasized or de-emphasized by clothing item
df234['body_part1'] = df234['branddescription_description_combine'].str.extract(r'(shoulders?|arms?|back|cleavage|waistline|rear|legs|calves|neck|butts?)')

## Niche products (organic material, local, sustainable, artists, poc, latinx)
df234['artisinal1'] = df234['branddescription_description_combine'].str.extract(r'(local?|organi[.|c]?\w{1,3}?|sustain?|sustainable?|artists?|poc|latinx?)')

## Patterm of clothing item
df234['patterns1'] = df234['branddescription_description_combine'].str.extract(r'(floral?|paisley?|polka?.?\s?dot?|\
herringbo?|herringbone?|stripe?|stripes?|plaid)')

## Fabric of clothing item
df234['fabric1'] = df234['branddescription_description_combine'].str.extract(r'(cotton|faux?.?\s?\w{1,7}?|leather|polyest?|polyester?|silk|wool|rayon|elastic|spandex)')

## Occasion associated with clothing item
df234['occasion1'] =  df234['branddescription_description_combine'].str.extract(r'(beach[e]?[s]?|city|business|formal|casual|professional|interview|wedding|dinner|date|dating\
|party|parties|baptism|funeral|cocktail|bar|night)')

## Description of bottom clothing items
df234['description_bottom1'] = df234['branddescription_description_combine'].str.extract(r'\b(pant[s]?|bottom[s]?|jean[s]?|leg|legging|jogger|sweatpant[s]?|trouser[s]?|short[s]?|skirt[s]?|culotte[s]?)\b')

## Description of one-piece clothing items
df234['one_piece_pattern1'] = df234['branddescription_description_combine'].str.extract(r'\b(jumpsuit[s]?|romper[s]?|dress|playsuit[s]?|onepiece|one-piece|unitard|gown[s]?|robe[s]?|cloak[s]?|bodysuit[s]?|swimsuit[s]?)\b')

## Descrption of shoes
df234['shoe_pattern1'] = df234['branddescription_description_combine'].str.extract(r'\b(shoe[s]?|sneaker[s]?|sandal[s]?|slipper[s]?|boot[s]?|heel[s]?|pump[s]?|flat[s]?|espadrille[s]?|slide[s]?|flip-flop[s]?|flip flop[s]?|flipflop[s]?|wedge[s]?\
    |loafer[s]?|oxford[s]?|mule[s]?|clog[s]?|chukka[s]?|chelsea[s]?|combat[s]?|moccasin[s]?|driver[s]?|derby[s]?|blucher[s]?)\b')

## Description of accessories
df234['accessory_pattern1'] = df234['branddescription_description_combine'].str.extract(r'\b(accessory|accessories|bag[s]?|backpack[s]?|tote[s]?|wristle[s]?t|handbag[s]?|card case[s]?|card holder[s]?|satchel[s]?|cross\s?body[s]?|wallet[s]?|pouch[s]?\
    |purse[s]?|clutch[s]?|bucket[s]?|pack[s]?|scarf|scarves|wrap|pashmina|shawl|stole|neckwear[s]?|kerchief|boa|bra[s]?|belt[s]?|bikini[s]?)\b')


## Description of top/upper clothing items
df234['branddescription_top1']= df234['branddescription_description_combine'].str.extract(r'\b(shirt[s]?|coat[s]?|blazer[s]?|tshirt[s]?|blouse[s]?|sweatshirt[s]?|tanks|camis|top[s]?|sweater[s]?|knit[s]?\
    |knitwear[s]?|sleeve[s]?|hoody|hoodies)\b')

## Gender associated with item type
df234['gender1'] = df234['branddescription_description_combine'].str.extract(r'(?P<gender1>\bm[a|e]n+?\b|\bwom[a|e]n+?\b|\bhim\b|\bher\b|\bhe\b|\bshe\b|\bhis\b)')


## Age associated with item type
df234['age1'] = df234['branddescription_description_combine'].str.extract(r'(?P<age1>\bchild+?|\bkids?\b|\binfants?\b|\badults?\b|\bteens?\b|\bteenagers?\b|\bbabys?\b|\bbabies\b|\badolescents?\b|\btoddlers?\b|\bchildre?)')

## fill any NaN results from creating the above custom features
df234['branddescription_description_combine'] = df234['branddescription_description_combine'].fillna("")

## fillna in the entire dataset
df234.fillna("",inplace=True)

# Top 50 brands
   
   Group by the top 50 brands that had the most products for sale. 

   Assign an ordered identifier number (0 - 49) for the top 50 brands and the identifier 99 for all other brands. 
   
   NOTE: Order of number assignment does not imply an ordinal importance.

In [61]:
## make brands all lower case 

df234['brand'] = df234['brand'].str.lower()
df234['branddescription_description_combine'] = df234['branddescription_description_combine'].str.lower()

 Create a sub-DataFrame as a group by on brands with a count of their available number of products for sale, sorting on the top 50 brands that had the most products for sale. An ordinal number is assigned to our top 50 brands in the section below:

In [62]:
# group by brand and count their number of products, sorting from the highest to the lowest amount of products
  # save top 50 
  
top50 = df234.groupby('brand')['product_id'].count()

top50 = top50.reset_index()  # save groupby as complete dataframe with both columns

top50 = top50.sort_values(by='product_id', ascending=False)

top50.head(50)
# save list of top 50 brands

top50 = top50.iloc[0:50]

top50['brand_number'] = top50['brand'].factorize()[0]

top50

brand  product_id  brand_number
186                      j.crew       13241             0
370                       vince       11044             1
150                       frame        9741             2
5             7 for all mankind        9656             3
268                  nili lotan        8128             4
293                       rails        7440             5
267                        nike        6541             6
348                      theory        5143             7
264                    nanushka        5050             8
352                  tory burch        5006             9
38                          alo        4613            10
51                   anine bing        4194            11
103                       chloé        3807            12
316       sarto by franco sarto        3740            13
313                 sam edelman        3253            14
0                                      3170            15
236              mansur gavriel        2896            16
333                       staud        2766            17
36                    allsaints        2684            18
182         intentionally blank        2582            19
7                        a.l.c.        2553            20
158                       ganni        2524            21
359                ulla johnson        2477            22
62   atm anthony thomas melillo        2384            23
207                    l'agence        2323            24
71              banana republic        2314            25
290                rachel comey        2081            26
254                        misa        2030            27
367              veronica beard        1975            28
372                     wandler        1971            29
309               saint laurent        1967            30
337                  studio 189        1956            31
202                      khaite        1949            32
61               astr the label        1942            33
295                     re/done        1934            34
183               isabel marant        1917            35
90                       by far        1882            36
218                      lemlem        1821            37
231          madeleine thompson        1780            38
163               gigi new york        1757            39
194                        joie        1706            40
323                simon miller        1670            41
193                  jimmy choo        1562            42
116             common projects        1550            43
87                brochu walker        1520            44
100               charles david        1518            45
318                         sea        1415            46
214             le 17 septembre        1405            47
288            proenza schouler        1372            48
122              cynthia rowley        1347            49

Save brand numbers to a dictionary and map them back to our main DataFrame.
We assign 99 as the number identifier to all other brands

### Assign number to top 50 brands by `product_id` count


In [63]:
## Assign numbers for top 50 brands based on product_id
    
top50dict = pd.Series(top50.brand_number.values, index=top50.brand).to_dict()

## Assign number to each of the top 50 brands

df234['brand_number'] = df234['brand']

df234['brand_number'] = df234['brand_number'].map(top50dict)

## Assign number to all other brands
df234['brand_number'].fillna(99, inplace=True)

In [64]:
## verify that all brand numbers are present ( 0 - 49 for the top50 brands, and 99 for all other brands)
df234['brand_number'].isnull().sum()

0

### Remove rows where all columns are NaN

In [65]:
df234.isnull().all(1).sum()  ## there are None

0

In [66]:
df234.index[df234.isnull().all(1)]  ## None

Int64Index([], dtype='int64')

# TF-IDF Vectorizer  Logistic Regression Model
  The TF-IDF vectorizer will assign a number vector to each product based on the frequency unique words used in the products description field, relative to presence of said words over all of the documents in our word corpus. Our word corpus comprised of roughly 268,000 product descriptions

Use sklearn TF-IDF and stopwords package to initiate a TF-IDF vectorizer.
Our vectorizer will look for naturally ocurring unigram through trigram occurrences of words. 

Parameters include: unigram through tri-gram expressions, minimum word occurrences set to 5 times in the corpus; 5,000 total features, and remove English stopwords from the corpus

In [67]:
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords

vectorizer = TfidfVectorizer(ngram_range=(1,3),
                             max_features = 5000,
                             min_df=5,
                             token_pattern=r'\b[a-zA-Z]{3,}\b',
                             max_df=0.4, stop_words=stopwords.words('english'))

Save combined-product-description column as our corpus, with all of string values within saved as a list.

Transform our corpus into a TF-IDF representation using the parameters established when creating our vectorizer (above).

In [68]:
# checking NaN values
df234["branddescription_description_combine"].isnull().sum()

0

In [69]:
# fill in NaN values
df234["branddescription_description_combine"]=df234["branddescription_description_combine"].fillna("")

In [70]:
## perform TFIDF Vectorizer

corpus = list(df234["branddescription_description_combine"].values)

X = vectorizer.fit_transform(corpus)

Use built in logistic regression in sklearn to train and predict which the brand using our TF-IDF vecorizer.

Perform a training / test data split by a ratio of 80/20 respectively, while stratifying our two data sets to ensure equal representation of all of the brands (top 50 + other brands).

In [71]:
from sklearn.linear_model import LogisticRegression


X   #<-- X takes the tf-idf vector as the values to regress on 
y = df234["brand_number"].values

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,
                                                    random_state = 0, stratify=y)

## Initiate and fit model
lr = LogisticRegression(max_iter=300)
lr.fit(X, y)

## Accuracy on baseline logistic regression

y_pred = lr.predict(X)

# calculate accuracy
np.mean(y_pred == y)

/Users/angela/opt/anaconda3/lib/python3.7/site-packages/sklearn/utils/__init__.py:806: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  return floored.astype(np.int)
/Users/angela/opt/anaconda3/lib/python3.7/site-packages/sklearn/utils/__init__.py:806: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your c

0.8666907326222711

Our Logistic Regression model produced a 86.67% accuracy rate. 

In [72]:
np.version.version

'1.20.2'